### PR TITLE
Add feature attribution

### PR DESCRIPTION
**Pull request by jackiehanyang**
### Description

[Describe what this change achieves]
UI changes:
- in anomaly result chart, show feature contribution of this anomaly with RectAnnotation component
- in feature break down chart, show expected value when there's an anomaly and current value when there's no anomaly as an additional <LineSeries> on the chart
- swap feature break down tab with anomaly occurrence tab to make it the default selected tab

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/299

[List any issues this PR will resolve]

### Check List

- [x] New functionality includes testing. Tested on localhost
  - [ ] All tests pass. will add more tests in the upcoming pr
- [ ] New functionality has been documented. 
- [x] Commits are signed per the DCO using `--signoff`
Screen Shot 2022-08-05 at 2 32 50 PM
Screen Shot 2022-08-05 at 1 42 56 PM

Screen Shot 2022-08-05 at 2 32 43 PM

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by sean-zheng-amazon — APPROVED** [review]
overall looks good. sync'ed offline, may need to make minor adjustments to improve consistency.

File: public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx [hunk 1]
```diff
@@ -278,9 +280,29 @@ export const FeatureChart = (props: FeatureChartProps) => {
...
-              );
+              )
+              if (featureTimeSeries.map(
+                (item: FeatureAggregationData) => {
+                  if(item.hasOwnProperty('expectedValue')) {
+                    componentList.push(
+                      <LineSeries
+                        id={"ExpectedValue"}
+                        name={"Expected Value"}
+                        color={"#0475a2"}
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx:291]: What is this color? Can you add a comment in the code about it somewhere? If supposed to be invisible, may need to find a way to always keep it consistent with the chart background, including if Dashboards is in dark mode or not

File: public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx [hunk 2]
```diff
@@ -257,13 +258,14 @@ export const FeatureChart = (props: FeatureChartProps) => {
           }
           {featureData.map(
             (featureTimeSeries: FeatureAggregationData[], index) => {
+              const componentList: any[] = [];
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx:261]: nit: can this be renamed to something like `timeSeriesList`?

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 3]
```diff
@@ -344,6 +348,59 @@ export const AnomalyDetailsChart = React.memo(
       zoomRange.endDate,
     ]);
 
+
+    const customAnomalyContributionTooltip = (details?: string) => {
+      const anomaly = details ? JSON.parse(details) : undefined;
+      const contributionData = get(anomaly, `features`, {})      
+      const componentList = [];
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:355]: nit: rename to something like `featureAttributionList`?

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 4]
```diff
@@ -344,6 +348,59 @@ export const AnomalyDetailsChart = React.memo(
       zoomRange.endDate,
     ]);
 
+
+    const customAnomalyContributionTooltip = (details?: string) => {
+      const anomaly = details ? JSON.parse(details) : undefined;
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:353]: this would include anomaly results with grade 0 too, right? Suggest to rename to `anomalyResults` for clarity

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 5]
```diff
@@ -344,6 +348,59 @@ export const AnomalyDetailsChart = React.memo(
...
+    };
+
+
+    const generateContributionAnomalyAnnotations = (
+      anomalies: AnomalyData[][]
+    ): any[][] => {
+      let annotations = [] as any[];
+
+      anomalies.forEach((anomalyTimeSeries: AnomalyData[]) => {
+        console.log("anomalyTimeSeries: " + JSON.stringify(anomalyTimeSeries))
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:387]: nit: will want to remove this log line

File: public/pages/AnomalyCharts/utils/constants.ts [hunk 6]
```diff
@@ -90,6 +92,7 @@ export const DEFAULT_ANOMALY_SUMMARY = {
   minConfidence: 0,
   maxConfidence: 0,
   lastAnomalyOccurrence: '-',
+  contributions: '-'
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/utils/constants.ts:95]: I'm confused about this part - are we showing feature contributions for all anomalies in the chart as a metric somehow? The above fields are all shown as metric-like headers above the anomaly chart.

File: public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx (hunk 1 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx:291]: It's a bluish color and shouldn't be invisible. I need to attend the UX office hour next Tuesday to get this color finalized.

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 4 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:353]: no it won't. it's already been filtered in `generateContributionAnomalyAnnotations`

File: public/pages/AnomalyCharts/utils/constants.ts (hunk 6 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/utils/constants.ts:95]: This is not used, forgot to clean it

File: public/pages/utils/anomalyResultUtils.ts [hunk 7]
```diff
@@ -575,6 +578,23 @@ export const parseBucketizedAnomalyResults = (result: any): Anomalies => {
...
+        forEach((attribution: any) => {
+          relevantAttributionMap[get(attribution, 'feature_id')] = get(attribution, 'data');
+        });
+        const contributions: { [key: string]: FeatureContributionData } = {}
+        for (const [key] of Object.entries(featureDataMap)) {
+          contributions[key] = {
+            name: featureDataMap[key],
+            attribution: relevantAttributionMap[key]
+          }
+        }
```
Comment (ohltyler) [inline public/pages/utils/anomalyResultUtils.ts:597]: nit: you can make this type-safe by adding `as FeatureContributionData` at the end.

File: public/models/interfaces.ts [hunk 8]
```diff
@@ -234,14 +234,23 @@ export type AnomalyData = {
...
 };
 
 export type FeatureAggregationData = {
   data: number;
+  name?: string;
   endTime: number;
   startTime: number;
   plotTime?: number;
+  attribution?: number;
+  expectedValue?: number;
```
Comment (ohltyler) [inline public/models/interfaces.ts:248]: is this also shown on the chart as part of this PR? Or just for saving the data for future UI improvements?

File: public/pages/utils/anomalyResultUtils.ts [hunk 9]
```diff
@@ -584,7 +604,23 @@ export const parseBucketizedAnomalyResults = (result: any): Anomalies => {
...
+
+        const featureBreakdownAttributionMap: {[key: string]: number} = {};
+
+        get(rawAnomaly, 'relevant_attribution', []).filter(element => get(element, 'feature_id') && get(element, 'data')).
+        forEach((attribution: any) => {
+          featureBreakdownAttributionMap[get(attribution, 'feature_id')] = toFixedNumberForAnomaly(get(attribution, 'data'));
+        });
+             
+        const expectedValueMap: {[key: string]: number} = {};
+        get(rawAnomaly, 'expected_values[0].value_list', []).forEach((expect: any) => {
```
Comment (ohltyler) [inline public/pages/utils/anomalyResultUtils.ts:620]: can the `forEach` arg be renamed and typesafe? For example, `expectedValue: number`?

File: public/models/interfaces.ts (hunk 8 above)
Comment (jackiehanyang) [reply public/models/interfaces.ts:248]: this also shown on the chart. See `LineSeries` rendering in FeatureChart.

File: public/pages/utils/anomalyResultUtils.ts [hunk 10]
```diff
@@ -674,13 +712,31 @@ export const parsePureAnomalies = (
...
+      forEach((attribution: any) => {
+        relevantAttributionMap[get(attribution, 'feature_id')] = toFixedNumberForAnomaly(get(attribution, 'data'));
+      });
+      const contributions: { [key: string]: FeatureContributionData } = {}
+      for (const [key] of Object.entries(featureDataMap)) {
+        contributions[key] = {
+          name: featureDataMap[key],
+          attribution: relevantAttributionMap[key]
+        }
+      }
```
Comment (ohltyler) [inline public/pages/utils/anomalyResultUtils.ts:731]: looks to be duplicate of lines 581-597 - can you combine into a helper fn? Also, prefer this version which enforces `toFixedNumberForAnomaly` on line 723

File: public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx (hunk 1 above)
Comment (ohltyler) [reply public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx:291]: Ok - is there a way for you to work with them sooner? Concern is that 8/9 is scheduled as the final sanity testing day - see timeline https://github.com/opensearch-project/opensearch-build/issues/2271

File: public/pages/DetectorResults/containers/AnomalyHistory.tsx [hunk 11]
```diff
@@ -666,6 +666,7 @@ export const AnomalyHistory = (props: AnomalyHistoryProps) => {
       } as Anomalies;
       anomalyResults.push(entityAnomaliesResult);
     });
+    console.log("allAnomalyResultResponses: " + JSON.stringify(anomalyResults))
```
Comment (ohltyler) [inline public/pages/DetectorResults/containers/AnomalyHistory.tsx:669]: nit: can eventually be removed

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 12]
```diff
@@ -531,7 +587,18 @@ export const AnomalyDetailsChart = React.memo(
...
+                    id="featureAttributionAnnotation"
+                    style={{
+                      stroke: CHART_COLORS.ANOMALY_GRADE_COLOR,
+                      strokeWidth: 1,
+                      opacity: 0.1,
+                      fill: CHART_COLORS.ANOMALY_GRADE_COLOR,
+                    }}
+                    renderTooltip={customAnomalyContributionTooltip}
+                  />   */}
+
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:601]: I'm assuming this will be uncommented in future change. I'm ok to leave this as-is for now.

**Review by ohltyler — APPROVED** [review]
Overall ok with the current changes - have a couple days buffer for finishing up. Thanks for adding this!

**Comment by ohltyler** [comment]
Also, if you can clean up the PR description and add some screenshots of the changes in the chart that would be great.

Or could rename this PR to specify that this is just adding the expected value and feature contribution to the anomaly result schema, and future PR to enable it in the charts.

**Comment by ohltyler** [comment]
There's also some integ test failures - I'm assuming it is related to switching the default anomaly history tabs
```
       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  plugins/anomaly-detection-dashboard      01:04        1        1        -        -        - │
  │    s-plugin/create_detector_spec.js                                                            │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  plugins/anomaly-detection-dashboard      01:07        8        8        -        -        - │
  │    s-plugin/dashboard_spec.js                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  plugins/anomaly-detection-dashboard      01:57        5        4        1        -        - │
  │    s-plugin/detector_configuration_spe                                                         │
  │    c.js                                                                                        │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  plugins/anomaly-detection-dashboard      01:35       11       11        -        -        - │
  │    s-plugin/detector_list_spec.js                                                              │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  plugins/anomaly-detection-dashboard      04:37        6        2        4        -        - │
  │    s-plugin/historical_analysis_spec.j                                                         │
  │    s                                                                                           │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  plugins/anomaly-detection-dashboard      02:31        6        4        2        -        - │
  │    s-plugin/overview_spec.js                                                                   │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  plugins/anomaly-detection-dashboard      01:24        3        -        1        -        2 │
  │    s-plugin/real_time_results_spec.js                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  plugins/anomaly-detection-dashboard      03:41        3        -        3        -        - │
  │    s-plugin/sample_detector_spec.js                                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

File: public/models/interfaces.ts [hunk 13]
```diff
@@ -234,14 +234,23 @@ export type AnomalyData = {
...
   entity?: EntityData[];
   features?: { [key: string]: FeatureAggregationData };
-  aggInterval?: string;
+  contributions?: { [key: string]: FeatureContributionData };
+  aggInterval?: string; 
 };
 
 export type FeatureAggregationData = {
   data: number;
+  name?: string;
```
Comment (amitgalitz) [inline public/models/interfaces.ts:243]: confused on this name property, did we not have this before at all?

File: public/pages/utils/anomalyResultUtils.ts (hunk 9 above)
Comment (jackiehanyang) [reply public/pages/utils/anomalyResultUtils.ts:620]: sure, will refactor this part later

File: public/pages/utils/anomalyResultUtils.ts (hunk 10 above)
Comment (jackiehanyang) [reply public/pages/utils/anomalyResultUtils.ts:731]: sure, will refactor this part later

File: public/models/interfaces.ts (hunk 13 above)
Comment (jackiehanyang) [reply public/models/interfaces.ts:243]: no, we are passing feature id around, not feature name

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 14]
```diff
@@ -351,17 +351,29 @@ export const AnomalyDetailsChart = React.memo(
...
-          <div>
-            {key}: {attribution} <br />
-          </div>
-        )
-      }
+      const contributionData = get(anomaly, `contributions`, [])  
 
+      const featureData = get(anomaly, `features`, {}) 
+      let featureAttributionList = [] as any[];
+      if (Array.isArray(contributionData)) {
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: when would contributions not be an array?

File: server/routes/ad.ts [hunk 15]
```diff
@@ -963,8 +963,9 @@ export default class AdService {
 
       const detectorResult: AnomalyResult[] = [];
       const featureResult: { [key: string]: FeatureResult[] } = {};
-
+      
       get(response, 'hits.hits', []).forEach((result: any) => {
+        console.log("result: " + JSON.stringify(result))
```
Comment (amitgalitz) [inline server/routes/ad.ts:968]: can be removed I assume

File: public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx (hunk 1 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/components/FeatureChart/FeatureChart.tsx:291]: sure, will reach out to them offline to draw an conclusion on this

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx [hunk 16]
```diff
@@ -391,7 +402,7 @@ export const AnomalyDetailsChart = React.memo(
               {
               coordinates: {
                 x0: anomaly.startTime,
-                x1: anomaly.endTime,
+                x1: anomaly.endTime + (anomaly.endTime - anomaly.startTime),
```
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:405]: what is this change for?

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 14 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: when it's getting response by calling search result api route in stead of get detector result api route

File: server/routes/ad.ts [hunk 17]
```diff
@@ -963,11 +963,15 @@ export default class AdService {
 
       const detectorResult: AnomalyResult[] = [];
       const featureResult: { [key: string]: FeatureResult[] } = {};
+      
       get(response, 'hits.hits', []).forEach((result: any) => {
+        console.log("result: " + JSON.stringify(result))
```
Comment (ohltyler) [inline server/routes/ad.ts:968]: can remove

File: server/routes/ad.ts [hunk 18]
```diff
@@ -995,10 +996,14 @@ export default class AdService {
           // to know feature data belongs to which anomaly result
           features: this.getFeatureData(result),
         });
+        
         result._source.feature_data.forEach((featureData: any) => {
           if (!featureResult[featureData.feature_id]) {
             featureResult[featureData.feature_id] = [];
           }
+          // if (result._source.anomaly_grade > 0) {
```
Comment (amitgalitz) [inline server/routes/ad.ts:1004]: can be removed also as its now taken care of in the new code below

File: server/routes/ad.ts [hunk 19]
```diff
@@ -963,11 +963,15 @@ export default class AdService {
...
       const featureResult: { [key: string]: FeatureResult[] } = {};
+      
       get(response, 'hits.hits', []).forEach((result: any) => {
+        console.log("result: " + JSON.stringify(result))
         detectorResult.push({
           startTime: result._source.data_start_time,
           endTime: result._source.data_end_time,
           plotTime: result._source.data_end_time,
+          contributions: result._source.anomaly_grade > 0 
+            ? result._source.relevant_attribution : {},
```
Comment (ohltyler) [inline server/routes/ad.ts:974]: shouldn't the default value be an empty array?

File: server/routes/ad.ts [hunk 20]
```diff
@@ -992,10 +996,14 @@ export default class AdService {
...
           features: this.getFeatureData(result),
         });
+        
         result._source.feature_data.forEach((featureData: any) => {
           if (!featureResult[featureData.feature_id]) {
             featureResult[featureData.feature_id] = [];
           }
+          // if (result._source.anomaly_grade > 0) {
+          //   this.getExpectedValue(result._source.expected_values[0].value_list, featureData.feature_id)
+          // }
```
Comment (ohltyler) [inline server/routes/ad.ts:1006]: can remove

File: server/routes/ad.ts [hunk 21]
```diff
@@ -963,8 +963,9 @@ export default class AdService {
 
       const detectorResult: AnomalyResult[] = [];
       const featureResult: { [key: string]: FeatureResult[] } = {};
-
+
```
Comment (amitgalitz) [inline server/routes/ad.ts:966]: nit: you might have done this already but you should run the linter on everything, could be in a different PR if you have another. https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/DEVELOPER_GUIDE.md#formatting

File: server/routes/ad.ts [hunk 22]
```diff
@@ -1004,6 +1012,9 @@ export default class AdService {
               featureData.data != null && featureData.data !== 'NaN'
                 ? toFixedNumberForAnomaly(Number.parseFloat(featureData.data))
                 : 0,
+            name: featureData.feature_name, 
+            expectedValue: result.anomaly_grade > 0 
+            ? this.getExpectedValue(result._source.expected_values[0].value_list, featureData.feature_id) : featureData.data
```
Comment (ohltyler) [inline server/routes/ad.ts:1016]: if `result._source.expected_values` is empty this will throw an NPE

File: server/routes/ad.ts [hunk 23]
```diff
@@ -1008,6 +1013,8 @@ export default class AdService {
                 ? toFixedNumberForAnomaly(Number.parseFloat(featureData.data))
                 : 0,
             name: featureData.feature_name, 
+            expectedValue: result.anomaly_grade > 0 
+            ? this.getExpectedValue(result._source.expected_values[0].value_list, featureData.feature_id) : featureData.data
```
Comment (amitgalitz) [inline server/routes/ad.ts:1016]: if the anomaly grade is at 0, will the expectedValue not still have some content? Might be misunderstandings whats in featureData.data

File: server/routes/ad.ts [hunk 24]
```diff
@@ -1127,8 +1138,20 @@ export default class AdService {
           featureData.data != null && featureData.data !== 'NaN'
             ? toFixedNumberForAnomaly(Number.parseFloat(featureData.data))
             : 0,
+        name: featureData.feature_name,
+        expectedValue: rawResult.anomaly_grade > 0 
+        ? this.getExpectedValue(rawResult._source.expected_values[0].value_list, featureData.feature_id)
```
Comment (ohltyler) [inline server/routes/ad.ts:1139]: Same as https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296/files/b8841161353efd11c94a81d846f20803c4160b6c..dd182b7122e2d632ca374f4e01f82a3f21c85df7#r939266382

File: server/routes/ad.ts (hunk 21 above)
Comment (jackiehanyang) [reply server/routes/ad.ts:966]: yeah I didn't know this. Will run it in the next pr

File: server/routes/ad.ts (hunk 19 above)
Comment (jackiehanyang) [reply server/routes/ad.ts:973]: the type is defined as `contributions: { [key: string]: FeatureContributionData };`

File: server/routes/ad.ts (hunk 22 above)
Comment (jackiehanyang) [reply server/routes/ad.ts:1013]: it's replying on the `result.anomaly_grade > 0` check. but I can certainly add more checks to make sure the page won't crash

File: server/routes/ad.ts (hunk 23 above)
Comment (jackiehanyang) [reply server/routes/ad.ts:1013]: if the anomaly grade is at 0, there's no expected value. The current value is the expected value

File: server/models/types.ts [hunk 25]
```diff
@@ -121,13 +121,21 @@ export type AnomalyResult = {
   confidence: number;
   entity?: Entity[];
   features?: { [key: string]: FeatureResult };
+  contributions: { [key: string]: FeatureContributionData };
```
Comment (ohltyler) [inline server/models/types.ts:124]: can make this optional

File: server/routes/ad.ts (hunk 23 above)
Comment (amitgalitz) [reply server/routes/ad.ts]: will featureData.data be empty than if no expected value, leading to no expected value?

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 16 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:405]: make the rectangle stay at the center of the start time and end time

File: server/models/types.ts (hunk 25 above)
Comment (jackiehanyang) [reply server/models/types.ts:124]: kinda did this on purpose, this is a pretty nested object. Making it optional makes the parsing process later difficult

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 14 above)
Comment (ohltyler) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: Not sure what you mean about get detector result api route. I see the parsing you've done in `getAnomalyResults()`. Where does the parsing happen where it is an array instead?

File: public/pages/utils/anomalyResultUtils.ts [hunk 26]
```diff
@@ -594,6 +630,8 @@ export const parseBucketizedAnomalyResults = (result: any): Anomalies => {
             startTime: get(rawAnomaly, 'data_start_time'),
             endTime: get(rawAnomaly, 'data_end_time'),
             plotTime: get(rawAnomaly, 'data_end_time'),
+            attribution: featureBreakdownAttributionMap[get(feature, 'feature_id')],
```
Comment (ohltyler) [inline public/pages/utils/anomalyResultUtils.ts:633]: can you make `attribution`/`contribution` naming consistent? From my understanding they're being used interchangeably here.

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 14 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: there's no parsing for that route because it's processed at receiving response stage.  See line [964](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296/files#diff-cc594889ac44cb130233969adf3da57a18f360a2a77c57ee273aad5f39d6d733R964) in ad.ts file, a type was defined and returned with response data

File: server/models/types.ts (hunk 25 above)
Comment (ohltyler) [reply server/models/types.ts:124]: Older anomaly results may be read by this plugin though, where `contributions` does not exist.

File: public/pages/utils/anomalyResultUtils.ts (hunk 26 above)
Comment (jackiehanyang) [reply public/pages/utils/anomalyResultUtils.ts:633]: sure, will improve the naming

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 14 above)
Comment (ohltyler) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: I'm still unsure when contributions would be an array. Even `AnomalyResult` defines it as a map [here](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296/files#diff-71f2b9246d36bd7c633e08d58e01d11cc08532db25bb0e00fd1efe2aad804751R116)

Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: it's from here - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/pages/DetectorResults/containers/AnomalyHistory.tsx#L664

**Review by amitgalitz — APPROVED** [review]
LGTM on my end, made sure to keep track of what needs to be refactored/cleaned up in next PR. Will check out if I need to review/approve again if you need to address more of Tyler's comments

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx (hunk 14 above)
Comment (jackiehanyang) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:358]: the data has been flattened before passing in - [https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296/files#diff-45683a46376000d3a5181f2[…]457590f0212f08b208877beaR602](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296/files#diff-45683a46376000d3a5181f2d2b68401f594dfe84457590f0212f08b208877beaR602)

**Review by ohltyler — APPROVED** [review]
Overall LGTM, thanks for making the changes! Let's follow up on https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/296#discussion_r939252880
